src/thriveopedia/Thriveopedia.cs

```diff
@@ -390,6 +392,25 @@ private void OnForwardPressed()
         ChangePage(pageFuture.Pop().PageName, true, false);
     }
 
+    private void OnSearchUpdated(string newText)
+    {
+        pageTree.Clear();
+        pageTree.CreateItem();
+
+        var pageList = allPages.Keys.ToList();
```

Comment:
Does this really need to create a copy of the list? Can't the keys collection be used below where ever this variable is used?

Reply:
The loop modifies the values of the dictionary. Even though the keys stay the same, I was getting a concurrent modification error without creating a copy.

Reply:
Ah I see. How do you then ensure that items come back when the search terms are cleared? I thought that `allPages` would be the absolute source of truth in terms of what pages exist and it shouldn't get modified.

Reply:
That's what I was hoping, but the issue is that you can't add an existing `TreeItem` to a tree, only create new ones with `CreateItem`. In order for the references to items in the tree to stay correct, I need to replace them with the newly created versions.

None of that would be an issue if you could set the visibility of a `TreeItem` instead of removing/replacing it completely, but you can't do that either.

Reply:
> None of that would be an issue if you could set the visibility of a TreeItem instead of removing/replacing it completely, but you can't do that either.

Well, Godot itself uses pretty much the [same approach](https://github.com/godotengine/godot/blob/3.5/editor/filesystem_dock.cpp#L1985) as here by recreating the tree.

Reply:
I see so this works correctly (pages are restored when the search input is cleared)? I was just confused as I thought allPages would be where the actual pages were stored, not just the tree items.

Reply:
`allPages` is a dictionary and the keys of that dictionary are the source of truth for what pages to recreate when the search text is updated.

Reply:
I guess I just can't wrap my head around this... because on line 405 the if skips putting some items into allPages there. But then I guess partially the old items are kept there? And that works even though clear has been called on the page tree meaning the tree items should all be deleted.